added service implementation 101 doc (#188)
deislabs/spiderlightning: docs/service-implementation-101.md

# Service Implementation 101
> valid for: 2022-09-02

Are you hoping to learn how to implement a new service for our existing capabilities (i.e., kv, lockd, mq, or pubsub)? You have come to the right place!

> Note: If you would like to develop a new capability outside of our provided interfaces, you will first have to create its WIT (WebAssembly Interface Types) file. Please, create an issue to start a design discussion, and make a PR with the `.wit` file itself prior to its' `slight` implementation.

> The DeisLabs Engineering Team has designed each of its' capabilities to be easily extensible, and some macros to take care of boilerplate code for you. In this tutorial, we will implement a new `pubsub` implementation — a local one with Mosquitto!

Reviewer: ```suggestion
The DeisLabs Engineering Team has designed each of its capabilities to be easily extensible, and some macros to take care of boilerplate code for you. In this tutorial, we will implement a new `pubsub` implementation — a local one with Mosquitto!
```

# Service Implementation 101
> valid for: 2022-09-02

Are you hoping to learn how to implement a new service for our existing capabilities (i.e., kv, lockd, mq, or pubsub)? You have come to the right place!

> > Note: If you would like to develop a new capability outside of our provided interfaces, you will first have to create its' WIT. Please, create an issue to start a design discussion, and make a PR with the `.wit` file itself prior to its' `slight` implementation.

Reviewer: ```suggestion
> Note: If you would like to develop a new capability outside of our provided interfaces, you will first have to create its WIT (WebAssembly Interface Types) file. Please, create an issue to start a design discussion, and make a PR with the `.wit` file itself prior to its' `slight` implementation.
```

> The DeisLabs Engineering Team has designed each of its' main capabilities to be easily extensible, and some macros to take care of boiler-plate code for you. In this tutorial, we will implement a new `pubsub` implementation — a local one with Mosquitto!

Reviewer: ```suggestion
The DeisLabs Engineering Team has designed each of its' capabilities to be easily extensible, and some macros to take care of boilerplate code for you. In this tutorial, we will implement a new `pubsub` implementation — a local one with Mosquitto!
```

## Getting Started

> Assuming that you are on the root of the SpiderLightning repository that you locally cloned, to get started, do:

Reviewer: ```suggestion
Assuming that you are on the root of the SpiderLightning repository that you locally cloned, to get started, create a new branch:
```

```sh

> git branch my_local_pubsub_implementation c73826a1d522bc3e1d2f0387481880edecd3e3d3

Reviewer: ```suggestion
git checkout -b my_local_pubsub_implementation
```

Author: I want this tutorial to work regardless of the state of the main branch~ That's why I'm branching off of a specific commit hash.

Reviewer: Won't that create a large PR when targeting main, or cause the developer to need to rebase to get started?

```

## Adding a New Dependency

To ease our development, will be making use of the [mosquitto-rs](https://crates.io/crates/mosquitto-rs) crate, and a couple of others.

At the end of our `Cargo.toml` in the `crates/pubsub` directory, add:

```toml
# pubsub.mosquitto deps
mosquitto-rs = { version = "0.4.0", features = ["default", "vendored-openssl"]}
futures = "0.3"
async-channel = "1.5"
```

Next up, inside `lib.rs` in the `crates/pubsub/src` directory, we'll have to create a new variant for the `PubImplementor` and `SubImplementor` enums. Like so:

```rs
#[derive(Debug, Clone)]
enum PubImplementor {
    ConfluentApacheKafka(PubConfluentApacheKafkaImplementor),
    Local(MosquittoImplementor)
};

// [..]

#[derive(Debug, Clone)]
enum SubImplementor {
    ConfluentApacheKafka(SubConfluentApacheKafkaImplementor),
    Local(MosquittoImplementor)
}
```

> Note: Our `Local` variants for both `enum`s hold a struct called `MosquittoImplementor` but we haven't implemented that yet.

> With that done, we now have to take care of the `new` function for both `PubImplementor`, and `SubImplementor`. Considering that the `new` function is what is responsible for mapping a string from a user's slightfile to a specific implementor, this is the time where we decide what string should refer to our local implementation — let's say `pubsub.mosquitto`. Now, we can go ahead and change the `new` functions, like so:

Reviewer: ```suggestion
With that done, we now have to take care of the `new` function for both `PubImplementor`, and `SubImplementor`. Considering that the `new` function is what is responsible for mapping a string from a user's slightfile to a specific implementor, this is the time where we decide what string should refer to our local implementation — let's say `pubsub.mosquitto`. Now, we can change the `new` functions, like so:
```

// -snip-

    fn sub_subscribe_to_topic(&mut self, self_: &Self::Sub, topic: Vec<&str>) -> Result<(), Error> {
        match &self_.sub_implementor {
            SubImplementor::ConfluentApacheKafka(si) => si.subscribe_to_topic(topic)?,
            SubImplementor::Local(si) => {
                si.subscribe_to_topic(topic.iter().map(|s| s.to_string()).collect::<Vec<String>>())?
            }
        }

// -snip-
    fn sub_poll_for_message(
        &mut self,
        self_: &Self::Sub,
        timeout_in_secs: u64,
    ) -> Result<Message, Error> {
        Ok(match &self_.sub_implementor {
            SubImplementor::ConfluentApacheKafka(si) => {
                si.poll_for_message(timeout_in_secs)
                    .map(|f| pubsub::Message {
                        key: f.0,
                        value: f.1,
                    })?
            }
            SubImplementor::Local(si) => {
                si.poll_for_message(timeout_in_secs)
                    .map(|f| pubsub::Message {
                        key: Some("batch".as_bytes().to_vec()),
                        value: Some(f.as_bytes().to_vec()),
                    })?
            }
        })
    }
```

> In here, all we are doing is adding calls to the functions implemented by the `MosquittoImplementor` under the `Local` variant of the `Pub/SubImplementor` enums, so that we are handling the case where a user provides a `toml` with `pubsub.mosquitto` and makes use of its' functionality.

Reviewer: ```suggestion
In here, all we are doing is adding calls to the functions implemented by the `MosquittoImplementor` under the `Local` variant of the `Pub/SubImplementor` enums, so that we are handling the case where a user provides a `toml` with `pubsub.mosquitto` and makes use of its functionality.
```

Next up, you'll have to make a change to the slight runner itself at: `slight/src/commands/run.rs`. That is, instead of:
```rs
const PUBSUB_HOST_IMPLEMENTORS: [&str; 1] = ["pubsub.confluent_apache_kafka"];
```

We want the `PUBSUB_HOST_IMPLEMENTORS` array to be:
```rs
const PUBSUB_HOST_IMPLEMENTORS: [&str; 2] = ["pubsub.confluent_apache_kafka", "pubsub.mosquitto"];
```

Now, you just need to create a slightfile, say, `mosquitto_slightfile.toml`, for your new implementor under their respective example crates — In this case, we want to create the exact same `mosquitto_slightfile.toml` for both the `examples/pubsub-consumer-demo` and `examples/pubsub-producer-demo`. It should look like this:
```toml
specversion = "0.1"
secret_store = "configs.azapp"

[[capability]]
name = "pubsub.mosquitto"
```

To finish off, make sure to add your example in the list of example runs under the `run-rust` Makefile rule — for our new pubsub implementor, it will look like so:
```Makefile
    RUST_LOG=$(LOG_LEVEL) $(SLIGHT) -c './examples/pubsub-consumer-demo/mosquitto_slightfile.toml' run -m ./examples/pubsub-consumer-demo/target/wasm32-wasi/release/pubsub-consumer-demo.wasm &

> 	RUST_LOG=$(LOG_LEVEL) $(SLIGHT) -c './examples/pubsub-producer-demo/mosquitto_slightfile.toml' run -m ./examples/pubsub-producer-demo/target/wasm32-wasi/release/pubsub-producer-demo.wasm

Reviewer: nit: indentation here is off.

Author: fixed ✅

```

> With this, you should have the basic understanding on how to get started developing capabilities for Slight — Now, all you are missing is just the implementation of the service itself! We are looking forward to your contributions 🚀

Reviewer: ```suggestion
With this, you should have the basic understanding on how to get started developing capabilities for Slight. Now, all you are missing is just the implementation of the service itself! We are looking forward to your contributions 🚀
```
